### PR TITLE
Improve gulp watch perf and add --nofetch option

### PR DESCRIPTION
**Pull request by nikhilkh**
Improves by reducing the set of files that we watch. Also, 'watch' was broken because of fetching behavior resulting an endless watch - change loop.

### CODE REVIEW

**Comment by nikhilkh** [comment]
@riknoll @dblotsky

File: gulpfile.js [hunk 1]
```diff
@@ -249,7 +252,14 @@ gulp.task("regen", ["jekyll"], function () {
 });
 
 gulp.task("fetch", function (done) {
-    exec("node", [bin("fetch_docs.js"), FETCH_CONFIG, FETCH_DIR], done);
+    if (!argv.nofetch) {
```
Comment (dblotsky) [inline gulpfile.js]: This could also just check if the fetch directory exists. If you want to re-fetch, just run `gulp clean`.

File: gulpfile.js [hunk 2]
```diff
@@ -236,7 +239,7 @@ gulp.task("serve", ["build"], function () {
     });
 });
 
-gulp.task("build", ["configs", "data", "styles", "plugins"], function (done) {
+gulp.task("build", [ "configs", "data", "styles", "plugins"], function (done) {
```
Comment (dblotsky) [inline gulpfile.js]: Extra space there.

File: gulpfile.js [hunk 3]
```diff
@@ -24,6 +24,8 @@ var uglify     = require("gulp-uglify");
 var envify     = require("envify");
 var htmllint   = require("gulp-htmllint");
 var crawler    = require("simplecrawler");
+var argv       = require('yargs').argv;
+var yaml       = require('js-yaml');
```
Comment (dblotsky) [inline gulpfile.js]: These aren't used in the file.

File: gulpfile.js [hunk 4]
```diff
@@ -24,6 +24,8 @@ var uglify     = require("gulp-uglify");
 var envify     = require("envify");
 var htmllint   = require("gulp-htmllint");
 var crawler    = require("simplecrawler");
+var argv       = require('yargs').argv;
```
Comment (dblotsky) [inline gulpfile.js]: This file uses `gutil.env` for args parsing.

File: gulpfile.js [hunk 5]
```diff
@@ -210,9 +213,9 @@ gulp.task("watch", ["serve"], function () {
     );
     gulp.watch(
         [
-            path.join(ROOT_DIR, "**", "*.yml"),
-            path.join(SOURCE_DIR, "**", "*.html"),
-            path.join(SOURCE_DIR, "**", "*.md"),
+            path.join(DATA_DIR, "**", "*.yml"),
+            path.join(SOURCE_DIR, "**", "*.html") + "!" + path.join(DOCS_DIR, "**"),
+            path.join(SOURCE_DIR, "**", "*.md") + "!" + path.join(DOCS_DIR, "**") ,
```
Comment (dblotsky) [inline gulpfile.js]: This won't regenerate the site if the docs change. Perhaps wrap this in `if (gutil.env.nodocs)`.

File: gulpfile.js [hunk 6]
```diff
@@ -210,9 +213,9 @@ gulp.task("watch", ["serve"], function () {
     );
     gulp.watch(
         [
-            path.join(ROOT_DIR, "**", "*.yml"),
-            path.join(SOURCE_DIR, "**", "*.html"),
-            path.join(SOURCE_DIR, "**", "*.md"),
+            path.join(DATA_DIR, "**", "*.yml"),
```
Comment (dblotsky) [inline gulpfile.js]: We want to regenerate the site if the configs change too, and `DATA_DIR` doesn't include `CONFIG_DIR`.

Comment (nikhilkh) [inline gulpfile.js]: Fair enough - I removed this change as it was not helping the perf.

**Comment by nikhilkh** [comment]
@dblotsky I made some updates - please take a look

File: gulpfile.js [hunk 7]
```diff
@@ -249,7 +243,13 @@ gulp.task("regen", ["jekyll"], function () {
 });
 
 gulp.task("fetch", function (done) {
-    exec("node", [bin("fetch_docs.js"), FETCH_CONFIG, FETCH_DIR], done);
+    if (!fs.existsSync(FETCH_DIR)) {
+        exec("node", [bin("fetch_docs.js"), FETCH_CONFIG, FETCH_DIR], done);
+    } else {
+        gutil.log(gutil.colors.yellow(
+            "Skipping fetching external docs. Run 'gulp clean' first to initiate another fetch."));
```
Comment (dblotsky) [inline gulpfile.js]: Nitpick: extraneous newline.

File: gulpfile.js [hunk 8]
```diff
@@ -280,7 +280,11 @@ gulp.task("defaults", function () {
 });
 
 gulp.task("toc", ["fetch"], function (done) {
-    exec("node", [bin("toc.js"), DOCS_DIR, DATA_DIR], done);
+    if(gutil.env.prod) {
```
Comment (dblotsky) [inline gulpfile.js]: Nitpick: space after `if`.

File: gulpfile.js [hunk 9]
```diff
@@ -193,14 +193,6 @@ gulp.task("watch", ["serve"], function () {
     );
     gulp.watch(
         [
-            path.join(DOCS_DIR, "**", "*.md"),
-            path.join(DOCS_DIR, "**", "*.html"),
-        ],
-        {interval: WATCH_INTERVAL},
-        ["configs", "data"]
```
Comment (dblotsky) [inline gulpfile.js]: Just making sure: this is to disable regeneration of toc/version data and configs while watching, correct?

File: gulpfile.js [hunk 10]
```diff
@@ -211,8 +203,10 @@ gulp.task("watch", ["serve"], function () {
     gulp.watch(
         [
             path.join(ROOT_DIR, "**", "*.yml"),
-            path.join(SOURCE_DIR, "**", "*.html"),
-            path.join(SOURCE_DIR, "**", "*.md"),
+            path.join(SOURCE_DIR, "**", "*.html") + "!" + path.join(DOCS_DIR, "**"),
+            path.join(SOURCE_DIR, "**", "*.md") + "!" + path.join(DOCS_DIR, "**") ,
```
Comment (dblotsky) [inline gulpfile.js]: This won't generate the site if the docs files change, and we want to have the site regenerate when a docs file changes. Please enclose these in `if (gutil.env.nodocs)`.

File: gulpfile.js [hunk 11]
```diff
@@ -211,8 +203,10 @@ gulp.task("watch", ["serve"], function () {
...
+            path.join(DOCS_DIR, "en", "dev", "**", "*.md") ,
+            path.join(DOCS_DIR, "en", "dev", "**", "*.html"),
```
Comment (dblotsky) [inline gulpfile.js]: I didn't realise that although the docs folder is excluded, the English folder is included explicitly. Could you please add a comment to draw attention to these two includes?

File: gulpfile.js (hunk 9 above)
Comment (nikhilkh) [inline gulpfile.js]: Yup!

**Pull request merged**